ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java

```diff
@@ -342,7 +343,12 @@ public static PathFilter getBootstrapDirectoryFilter(final FileSystem fs) {
 
   public static int handleException(boolean isReplication, Throwable e, String nonRecoverablePath,
                                     ReplicationMetricCollector metricCollector, String stageName, HiveConf conf){
-    int errorCode = ErrorMsg.getErrorMsg(e.getMessage()).getErrorCode();
+    int errorCode;
+    if (isReplication && e instanceof SnapshotException) {
+      errorCode = ErrorMsg.getErrorMsg("SNAPSHOT_ERROR").getErrorCode();
```

Comment:
why is this different? Can't we compare based on error msg

Reply:
Actually it should be like that but, in the ErrorMsg, it gets the error code based on the Exception message, but in case of Snapshot Exception, There can be a bunch of messages, Different for Nested Snapshot, Parent Snapshottable, No Snapshot Exists and couple of more cases. So Identifying all of such case and there corresponding Error Message won't be possible, So, I catch all SnapshotException and get one Error code for them.

Reply:
Is the actual error msg retained so that users can check that?

Reply:
Yes, It will be preserved. The entire stack trace is written here. https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java#L353

and the exception is already set above, example:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java#L141

Every code that calls this sets the exception before and calls this for error code to check if it is recoverable or non-recoverable.
This logic is just to get the error code, which is used to decide whether the error is recoverable or non-recoverable.

And the trace is also propagated back, example from `testFailureScenarios()`:

```
Caused by: org.apache.hadoop.hdfs.protocol.SnapshotException: Nested snapshottable directories not allowed: path=/testFailureScenariossource1/tablesource, the ancestor /testFailureScenariossource1 is already a snapshottable directory.
        at org.apache.hadoop.hdfs.server.namenode.snapshot.SnapshotManager.checkNestedSnapshottable(SnapshotManager.java:174)
        at org.apache.hadoop.hdfs.server.namenode.snapshot.SnapshotManager.setSnapshottable(SnapshotManager.java:189)
        at org.apache.hadoop.hdfs.server.namenode.FSDirSnapshotOp.allowSnapshot(FSDirSnapshotOp.java:62)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.allowSnapshot(FSNamesystem.java:6366)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.allowSnapshot(NameNodeRpcServer.java:1842)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.allowSnapshot(ClientNamenodeProtocolServerSideTranslatorPB.java:1211)
        at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:523)

```